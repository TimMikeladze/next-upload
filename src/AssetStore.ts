import type Keyv from 'keyv';
import { Asset, NextUploadAssetStore } from './types';

export class AssetStore implements NextUploadAssetStore {
  private keyv: Keyv;

  constructor(keyv: Keyv) {
    this.keyv = keyv;
  }

  async upsert(args: Asset, ttl: number): Promise<Asset> {
    await this.keyv.set(args.id, args, ttl);
    return args;
  }

  async delete(id: string): Promise<void> {
    await this.keyv.delete(id);
  }

  async find(id: string): Promise<Asset | undefined> {
    return this.keyv.get(id);
  }

  iterator() {
    return this.keyv.iterator();
  }
}
