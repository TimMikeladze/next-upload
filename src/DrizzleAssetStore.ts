import {
  pgTable,
  jsonb,
  integer,
  timestamp,
  varchar,
} from 'drizzle-orm/pg-core';
import { drizzle } from 'drizzle-orm/postgres-js';
import { eq } from 'drizzle-orm';
import { nanoid } from 'nanoid';
import { Asset, NextUploadAssetStore } from './types';
import { NextUpload } from './NextUpload';

const createdAt = timestamp(`createdAt`, {
  withTimezone: true,
})
  .notNull()
  .defaultNow();

const updatedAt = timestamp(`updatedAt`, {
  withTimezone: true,
})
  .notNull()
  .defaultNow();

const assetsTable = pgTable(`next_upload_assets`, {
  createdAt,
  updatedAt,
  id: varchar(`id`).primaryKey(),
  data: jsonb(`data`).notNull(),
  expires: integer(`expires`),
});

export class DrizzlePostgresAssetStore implements NextUploadAssetStore {
  public static tables = {
    assetsTable,
  };

  private db: ReturnType<typeof drizzle>;

  constructor(db: ReturnType<typeof drizzle>) {
    this.db = db;
  }

  async all(): Promise<Asset[]> {
    const rows = await this.db
      .select()
      .from(DrizzlePostgresAssetStore.tables.assetsTable);

    return rows.map((row) => ({
      ...(row.data as Asset),
      ...row,
    }));
  }

  async upsert(args: Asset, ttl: number): Promise<Asset> {
    const found = await this.find(args.id);

    const expires = NextUpload.calculateExpires(ttl);

    // eslint-disable-next-line no-param-reassign
    delete args.expires;

    if (found) {
      if (NextUpload.isExpired(found)) {
        await this.delete(args.id);
        throw new Error(`Asset expired and was deleted`);
      }
      const rows = await this.db
        .update(DrizzlePostgresAssetStore.tables.assetsTable)
        .set({
          data: args,
          expires,
        })
        .where(eq(DrizzlePostgresAssetStore.tables.assetsTable.id, args.id))
        .returning();

      return {
        ...(rows?.[0].data as Asset),
        expires: rows?.[0].expires,
        id: rows?.[0].id,
        createdAt: rows?.[0].createdAt,
        updatedAt: rows?.[0].updatedAt,
      };
    }

    const rows = await this.db
      .insert(DrizzlePostgresAssetStore.tables.assetsTable)
      .values({
        id: args?.id || nanoid(),
        data: args,
      })
      .returning();

    return {
      ...(rows?.[0].data as Asset),
      expires: rows?.[0].expires,
      id: rows?.[0].id,
      createdAt: rows?.[0].createdAt,
      updatedAt: rows?.[0].updatedAt,
    };
  }

  async delete(id: string): Promise<void> {
    await this.db
      .delete(DrizzlePostgresAssetStore.tables.assetsTable)
      .where(eq(DrizzlePostgresAssetStore.tables.assetsTable.id, id));
  }

  async find(id: string): Promise<Asset | undefined> {
    const rows = await this.db
      .select()
      .from(DrizzlePostgresAssetStore.tables.assetsTable)
      .where(eq(DrizzlePostgresAssetStore.tables.assetsTable.id, id));

    if (rows?.[0]) {
      if (NextUpload.isExpired(rows?.[0])) {
        await this.delete(id);
        throw new Error(`Asset expired and was deleted`);
      }
    }

    return {
      ...(rows?.[0].data as Asset),
      expires: rows?.[0].expires,
      id: rows?.[0].id,
      createdAt: rows?.[0].createdAt,
      updatedAt: rows?.[0].updatedAt,
    };
  }
}
