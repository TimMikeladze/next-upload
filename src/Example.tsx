import './Example.css';

import React from 'react';

export type ExampleProps = {
  text?: String;
};

export function Example(props: ExampleProps) {
  const [count, setCount] = React.useState(0);
  return (
    <button
      onClick={() => setCount(count + 1)}
      type="button"
      id="example-button"
    >
      {`${props.text} ${count}`}
    </button>
  );
}
