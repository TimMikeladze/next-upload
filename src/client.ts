export * from './types';
export * from './getSignedUrl';
export * from './uploadToSignedUrl';
export * from './upload';
