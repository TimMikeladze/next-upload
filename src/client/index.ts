export * from '../types';
export * from './generatePresignedPostPolicy';
export * from './uploadToPresignedUrl';
export * from './upload';
export * from './getAsset';
