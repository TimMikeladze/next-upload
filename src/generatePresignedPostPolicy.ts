import {
  GeneratePresignedPostPolicyOptions,
  HandlerAction,
  NextUploadClientConfig,
  SignedPostPolicy,
} from './types';

export const generatePresignedPostPolicy = async (
  options: GeneratePresignedPostPolicyOptions,
  config: NextUploadClientConfig
): Promise<SignedPostPolicy> => {
  const api = config.api || `/upload`;
  const res = await fetch(api, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...options.requestInit?.headers,
    },
    body: JSON.stringify({
      action: HandlerAction.generatePresignedPostPolicy,
      args: options.args,
      ...options.requestInit?.body,
    }),
    ...options.requestInit,
  });

  const json = await res.json();

  if (!res.ok) {
    throw new Error(json.error);
  }

  return json;
};
