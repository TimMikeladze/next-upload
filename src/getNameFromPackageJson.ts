import { readFileSync } from 'fs';
import { resolve } from 'path';

export const getNameFromPackageJson = () => {
  const packageJson = JSON.parse(
    readFileSync(resolve(process.cwd(), 'package.json'), 'utf-8')
  );
  return packageJson.name;
};
