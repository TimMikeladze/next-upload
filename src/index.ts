export * from './types';
export * from './NextUpload';
export * from './store/keyv/KeyvAssetStore';
export * from './store/drizzle/pg/DrizzlePgAssetStore';
export * from './store/drizzle/pg/DrizzlePgSchema';
export * from './store/drizzle/mysql/DrizzleMysqlAssetStore';
