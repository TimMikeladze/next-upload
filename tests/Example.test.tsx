import React from 'react';
import renderer from 'react-test-renderer';
import { expect, it } from 'vitest';
import { Example } from '../src';

it('renders correctly', () => {
  const tree = renderer
    .create(<Example text="Clicked this many times" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
