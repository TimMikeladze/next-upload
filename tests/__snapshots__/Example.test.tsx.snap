// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<button
  id="example-button"
  onClick={[Function]}
  type="button"
>
  Clicked this many times 0
</button>
`;
